File: App022_Journal/app/src/main/java/com/kes/app022_journal/activities/JournalListActivity.java
package com.kes.app022_journal.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.storage.StorageReference;
import com.kes.app022_journal.MainActivity;
import com.kes.app022_journal.R;
import com.kes.app022_journal.adapters.JournalRecyclerAdapter;
import com.kes.app022_journal.models.CurrentUser;
import com.kes.app022_journal.models.Journal;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class JournalListActivity extends AppCompatActivity {

    // Widgets
    private RecyclerView recyclerView;
    private TextView noPostsView;
    MaterialToolbar toolbar;

    // Auth
    private FirebaseAuth firebaseAuth;
    private FirebaseAuth.AuthStateListener authStateListener;
    private FirebaseUser firebaseUser;

    // Database
    private FirebaseFirestore db = FirebaseFirestore.getInstance();
    private CollectionReference journalsDB = db.collection("Journals");
    private StorageReference storageReference;
    private List<Journal> posts;

    private JournalRecyclerAdapter journalRecyclerAdapter;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_journal_list);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        // Firebase Auth
        firebaseAuth = FirebaseAuth.getInstance();
        firebaseUser = firebaseAuth.getCurrentUser();

        // Widgets
        noPostsView = findViewById(R.id.list_no_posts);
        recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));

        // Posts
        posts = new ArrayList<>();

        // Toolbar menu
        toolbar = findViewById(R.id.appbar);
        toolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                return HandleItemClick(item);

            }
        });
    }

    private boolean HandleItemClick(MenuItem item) {
        if (item.getItemId() == R.id.menu_action_add) {
            // Redirect to add journal activity
            if (firebaseUser != null && firebaseAuth != null) {
                startActivity(new Intent(getApplicationContext(), AddJournalActivity.class));
                return true;
            }
        } else if (item.getItemId() == R.id.menu_action_sign_out) {
            // Logout and redirect to main activity
            if (firebaseUser != null && firebaseAuth != null) {
                firebaseAuth.signOut();
                startActivity(new Intent(getApplicationContext(), MainActivity.class));
                return true;
            }
        }

        return false;
    }

    @Override
    protected void onStart() {
        super.onStart();

        journalsDB.whereEqualTo("userID", firebaseUser.getUid())
                .get()
                .addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        if (queryDocumentSnapshots.isEmpty()) {
                            noPostsView.setVisibility(TextView.VISIBLE);
                            return;
                        }

                        // Add retrieved documents to array of posts
                        for (QueryDocumentSnapshot document : queryDocumentSnapshots) {
                                posts.add(document.toObject(Journal.class));
                        }

                        // RecyclerView
                        journalRecyclerAdapter = new JournalRecyclerAdapter(
                                getApplicationContext(),
                                posts
                        );
                        recyclerView.setAdapter(journalRecyclerAdapter);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Toast.makeText(JournalListActivity.this, "Failed to load Posts", Toast.LENGTH_SHORT).show();
                        Log.e("ERROR", Objects.requireNonNull(e.getMessage()));
                    }
                });

    }
}

File: App040_Kt_Journal/app/src/main/java/com/kes/app040_kt_journal/activities/JournalListActivity.kt
package com.kes.app040_kt_journal.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.PopupMenu
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.Firebase
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.auth
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.storage.StorageReference
import com.kes.app040_kt_journal.MainActivity
import com.kes.app040_kt_journal.R
import com.kes.app040_kt_journal.adapters.JournalRecyclerAdapter
import com.kes.app040_kt_journal.databinding.ActivityJournalListBinding
import com.kes.app040_kt_journal.models.Journal

class JournalListActivity : AppCompatActivity() {

    private lateinit var binding: ActivityJournalListBinding

    // Firebase
    private lateinit var auth: FirebaseAuth
    private lateinit var user: FirebaseUser
    private lateinit var storage: StorageReference
    private var db: FirebaseFirestore = FirebaseFirestore.getInstance()
    private var collection: CollectionReference = db.collection("KtJournal")

    private lateinit var journalList: MutableList<Journal>
    private lateinit var adapter: JournalRecyclerAdapter

    private lateinit var noPostsMessage: CardView


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binding = DataBindingUtil.setContentView(this, R.layout.activity_journal_list)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        // Firebase Auth
        auth = Firebase.auth
        user = auth.currentUser!!

        // RecyclerView
        binding.recyclerView.setHasFixedSize(true)
        binding.recyclerView.layoutManager = LinearLayoutManager(this)
        noPostsMessage = binding.noPostsCard

        // Journals
        journalList = mutableListOf()
        binding.fabAdd.setOnClickListener { createJournal() }
        binding.buttonMenu.setOnClickListener {
            showMenu(it)
        }



    }

    override fun onStart() {
        super.onStart()

        collection.whereEqualTo("userID", user.uid).get()
            .addOnSuccessListener {
                if (!it.isEmpty) {
                    it.forEach { document ->
                        val journal = document.toObject(Journal::class.java)
                        journalList.add(journal)
                    }
                    adapter = JournalRecyclerAdapter(this, journalList)
                    binding.recyclerView.adapter = adapter
                    adapter.notifyItemRangeChanged(0, journalList.size)
                } else {
                    noPostsMessage.visibility = View.VISIBLE
                }
            }
            .addOnFailureListener {
                Toast.makeText(this, "Could not get journals!", Toast.LENGTH_SHORT).show()
                noPostsMessage.visibility = View.VISIBLE
                binding.noPostsCardText.text = it.toString()
                Log.w("TAG_E", it.toString())
            }



    }

    private fun createJournal() {
        if (user.isAnonymous) {
            Toast.makeText(this, "Sign-In required", Toast.LENGTH_SHORT).show()
            return
        }
        val intent = Intent(this, AddJournalActivity::class.java)
        startActivity(intent)
    }

    private fun showMenu(view: View) {
        val popup = PopupMenu(this, view)
//        val inflater: MenuInflater = popup.menuInflater

        popup.apply {
            setOnMenuItemClickListener { onMenuItemClick(it) }
            inflate(R.menu.popup)
            show()
        }
    }

    private fun onMenuItemClick(item: MenuItem) : Boolean {
        return when (item.itemId) {
            R.id.popup_action_sign_out -> {
                signOut()
                true
            }
            else -> false
        }
    }

    private fun signOut() {
        auth.signOut()
        val intent = Intent(this, MainActivity::class.java)
        startActivity(intent)
    }


}